kotest-property/src/commonMain/kotlin/io/kotest/property/state/state.kt

```diff
@@ -0,0 +1,54 @@
+package io.kotest.property.state
+
+import io.kotest.property.RandomSource
+
+/**
+ * A function that is applied against the current state to mutate and/or test.
+ */
+interface Action<T> {
+
+   /**
+    * A predicate which decides if this action should be applied.
+    *
+    * For example, if you were writing a stateful test for a stack, you would want
+    * to skip 'pop' operations if the stack was empty.
+    *
+    * @return true if this action should be applied or false if the action should be skipped.
+    */
+   suspend fun precondition(state: T): Boolean
+
+   /**
+    * Invokes the next transition to the state.
+    *
+    * This function should contain your test assertions.
+    */
+   suspend fun run(state: T, rs: RandomSource)
+
+   companion object {
+
+      /**
+       * Creates a new [Action] that is always applied (precondition is set to constant true).
+       */
+      operator fun <T> invoke(f: (T, RandomSource) -> Unit): Action<T> {
```

Comment:
I wonder if we can instead use a `GenerateArbitraryBuilderContext` on this so we actually discourage users from calling `.next(rs)`.. To my experience that habit can be detrimental to the project health, especially with a team of many.. That is because the temptation of just calling `.next()` without random seed is real, and can easily creep. I'm currently thinking of writing a ktlint rule to ban the use of `.next()` and `.single()` throughout our (in my company, not kotest) codebase.

my suggestion would be:
```kotlin
      operator fun <T> invoke(f: suspend GenerateArbitraryBuilderContext.(T) -> Unit): Action<T> {
         return object : Action<T> {
            override suspend fun precondition(state: T): Boolean = true
            override suspend fun run(state: T, rs: RandomSource) {
               generateArbitrary { f(state) }.generate(rs).first()
            }
         }
      }
```

Reply:
I looked at using an ArbitraryBuilderContext but won't it capture the continuation so we can't use suspend functions inside the action? So it means we can't have an action to delay say, useful for testing cache expiry.

Reply:
There's another one Sam, `generateArbitrary { }` the suspend version

Reply:
Can I use delay etc inside that one ?

Reply:
Can I use delay etc inside that one ?

Reply:
`generateArbitrary` is not using restricted continuation and is designed to be used in conjunction with suspend functions in mind. we have several of those which setup dynamodb state etc in our codebase with arbitrary data.

Reply:
and yes of course, you can use delay 💯

Reply:
Oh awesome, then that's the one to use, I'll update.